racket/src/thread/thread.rkt

```diff
@@ -556,13 +557,16 @@
   ;; awoken or there was already mail.
   (define t (current-thread))
   (let loop ()
-    (cond
-      [(is-mail? t)
-       (atomically
-        (dequeue-mail! t))]
-      [else
-       ((thread-internal-suspend! t #f void block))
-       (loop)])))
+    ((atomically
+      (cond
+        [(is-mail? t)
+         (let ([msg (dequeue-mail! t)])
+           (lambda () msg))]
+        [else
+         (let ([f (thread-internal-suspend! t #f void block)])
+           (lambda ()
+             (f)
+             (loop)))])))))
```

Comment:
I agree that atomic mode is needed – and then this covers all of the work of the first atomic block in `thread-receive`, so I'll remove that first one